test/distributed/checkpoint/e2e/test_fine_tuning.py

```diff
@@ -0,0 +1,192 @@
+        self.layer2 = nn.Linear(DIM, DIM)
+        self.layer3 = nn.Linear(DIM, DIM)
+        self.relu = nn.ReLU()
+
+    def forward(self, batch):
+        x = self.relu(self.layer1(batch))
+        x = self.relu(self.layer2(x))
+        x = self.relu(self.layer3(x))
+        return x
+
+
+class FineTuningModel(nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.pretrain = PreTrainedModel()
+        for p in self.pretrain.parameters():
+            p.requires_grad = False
+
+        self.layer1 = nn.Linear(DIM, DIM)
+        self.layer2 = nn.Linear(DIM, DIM)
+        self.layer3 = nn.Linear(DIM, DIM)
+        self.relu = nn.ReLU()
+
+    def forward(self, batch):
+        x = self.relu(self.pretrain(batch))
+        x = self.relu(self.layer1(x))
+        x = self.relu(self.layer2(x))
+        x = self.relu(self.layer3(x))
+        return x
+
+
+class TestFineTuning(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(4, torch.cuda.device_count())
+
+    def pretrain(self, pretrain_dir: str) -> None:
+        model = PreTrainedModel().cuda()
+        model = FSDP(model)
+        optim = torch.optim.Adam(model.parameters(), lr=1e-3)
+
+        # Trainining
+        for i in range(3):
+            batch = torch.rand(32, DIM, device="cuda")
+            loss = model(batch).sum()
+            loss.backward()
+            optim.step()
+            optim.zero_grad()
+
+        # Save state_dict
+        model_state_dict, optim_state_dict = get_state_dict(model, optimizers=optim)
+        saved_state_dict = {"model": model_state_dict, "optim": optim_state_dict}
+        dist_cp.save_state_dict(
+            state_dict=saved_state_dict,
+            storage_writer=dist_cp.FileSystemWriter(pretrain_dir),
+        )
+
+    def finetune(self, pretrain_dir: str, finetune_dir: str) -> None:
+        model = FineTuningModel().cuda()
+        # TODO: make the parallelism more complicated, e.g., using 2D + DDP.
+        model = FSDP(model, use_orig_params=True)
+        optim = torch.optim.Adam(model.parameters(), lr=1e-3)
+
+        # Simulate that the fine tuning restart after 3 iterations
+        for i in range(2):
+            # Load pretrain submodules checkpoint
+            pretrain_state_dict, _ = get_state_dict(
+                model,
+                submodules={model.pretrain},
+                options=StateDictOptions(keep_submodule_prefixes=False)
+            )
+            dist_cp.load_state_dict(
+                {"model": pretrain_state_dict},
+                storage_reader=dist_cp.FileSystemReader(pretrain_dir),
+            )
+            set_state_dict(
+                model,
+                model_state_dict={model.pretrain: pretrain_state_dict},
+                options=StateDictOptions(strict=False),
+            )
+
+            try:
+                # Load training submodules checkpoint
+                model_state_dict, optim_state_dict = get_state_dict(
```

Comment:
This might just be me, but this sort of pattern of having to get the state_dict and then pass it into DCP is a bit confusing, because it almost reads as if I'm saving a state_dict here, but I'm actually just creating a holder and DCP will inplace load into it. 

I know that DCP has to do this, but could we maybe support a higher-level `load_state_dict` API that manages this for users?

Reply:
That's why I changed the name to `get_state_dict`, which means getting the current state_dict. It can be used for saving but can also be used for loading. The full story for loading is getting the current state dict to represent the current sharding of the model (hence `get_state_dict`) and DCP load the tensor data based on the current sharding (hence `load_state_dict`).

It is not easy to merge this get_state_dict to DCP load_state_dict. For example, we may want to have a centralize checkpoint for all different fine tuning model. So each model will require to read 2 different checkpoints, like this example. Having only one `load_state_dict` will make things much harder to achieve.